File: Top_100_Liked_Questions/253._Meeting_Rooms_II/solution.js
/**
 * @param {number[][]} intervals
 * @return {number}
 */
var minMeetingRooms = function(intervals) {    
    // sort intervals by first the starting time and then the end time
    intervals.sort((a, b) => a[0] - b[0] || a[1] - b[1]);
    
    // keep track of the endtimes in minHeap
    let endTimes = new MinHeap();
    
    for (let [s,e] of intervals) {
        // if the starting time is before the end time, push the new endtime into the heap
        if (endTimes.size() && s >= endTimes.peek()) {
            endTimes.pop();
        }
        endTimes.push(e);
    }
    
    return endTimes.size();
};

function MinHeap(arr=[]) {
    this.heap = [];
    this.heapify(arr);
}

MinHeap.prototype.peek = function() {
    return this.heap[0];
}

MinHeap.prototype.size = function(i) {
    return this.heap.length;
}

MinHeap.prototype.push = function(n) {
    this.heap.push(n);
    this.siftUp(this.heap.length-1);
}

MinHeap.prototype.pop = function() {
    if (this.heap.length == 1) {
        return this.heap.pop();
    }
    
    let popVal = this.heap[0];
    this.heap[0] = this.heap.pop();
    this.siftDown(0);
    return popVal;
}

MinHeap.prototype.heapify = function(arr) {
    this.heap = arr;
    for (let i=this.heap.length-1; i>-1; i--) {
        this.siftDown(i);
    }
}

MinHeap.prototype.siftUp = function(i) {
    while (i > 0) {
        let parIdx = this.getParent(i);
        if (this.heap[i] < this.heap[parIdx]) {
            this.swap(i, parIdx);
            i = parIdx;
        } else {
            break;
        }
    }
}

MinHeap.prototype.siftDown = function(i) {
    let leng = this.heap.length - 1;
    while (i < leng) {
        let [leftIdx, rightIdx] = this.getChildren(i);
        
        if (leftIdx > leng) {
            break;
        }
        
        let minChildIdx = (rightIdx <= leng && this.heap[rightIdx] < this.heap[leftIdx]) ? rightIdx : leftIdx;
        
        if (this.heap[i] <= this.heap[minChildIdx]) {
            break;
        }
        
        this.swap(i, minChildIdx);
        i = minChildIdx;
    }
}

MinHeap.prototype.getParent = function(i) {
    return Math.floor((i - 1) / 2);
}

MinHeap.prototype.getChildren = function(i) {
    return [i*2+1, i*2+2];
}

MinHeap.prototype.swap = function(i, j) {
    [this.heap[i], this.heap[j]] = [this.heap[j], this.heap[i]];
}

File: Top_100_Liked_Questions/253._Meeting_Rooms_II/solution.py
class Solution:
    def minMeetingRooms(self, intervals: List[List[int]]) -> int:
        intervals.sort(key=lambda x: (x[0], x[1]))
        
        endTimes = minHeap()
        
        for s,e in intervals:
            if endTimes.size() and s >= endTimes.peek():
                endTimes.pop()
                
            endTimes.push(e)
        
        return endTimes.size()
        
        
    
class minHeap:
    def __init__(self, arr=[]):
        self.heap = []
        self.heapify(arr)
    
    
    def heapify(self, arr):
        self.heap = arr
        for i in range(len(self.heap)-1, -1, -1):
            self.siftDown(i)
    
    
    def push(self, n):
        self.heap.append(n)
        self.siftUp(len(self.heap)-1)
    
    
    def pop(self):
        if len(self.heap) == 1:
            return self.heap.pop()
        
        popVal = self.heap[0]
        self.heap[0] = self.heap.pop()
        self.siftDown(0)
        return popVal
    
    
    def peek(self):
        return self.heap[0]
    
    
    def siftUp(self, i):
        while i > 0:
            parentIdx = self.getParentIdx(i)
            if self.heap[i] >= self.heap[parentIdx]:
                break
            
            self.swap(i, parentIdx)
            i = parentIdx
    
    
    def siftDown(self, i):
        leng = len(self.heap) - 1
        while i < leng:
            leftIdx, rightIdx = self.getChildrenIdx(i)
            
            if leftIdx > leng:
                break
            
            minChildIdx = rightIdx if rightIdx <= leng and self.heap[rightIdx] < self.heap[leftIdx] else leftIdx
            
            if self.heap[i] <= self.heap[minChildIdx]:
                break
            
            self.swap(i, minChildIdx)
            i = minChildIdx
    
    
    def getParentIdx(self, i):
        return (i-1)//2
    
    
    def getChildrenIdx(self, i):
        return i*2+1, i*2+2
    
    
    def size(self):
        return len(self.heap)
    
    
    def swap(self, i, j):
        self.heap[i], self.heap[j] = self.heap[j], self.heap[i]
    